### PR TITLE
Actual async slide loading

### LINKED ISSUE #812: Slide Reel loading lag
### What Happened?

Slide reels are loaded all at once causing incredibly long load times instead of being loaded asynchronous.

### What was supposed to happen?

_No response_

### Platform

None

### Mods

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by JohnCorby**
## Bug fixes

- Slides are now loaded in the background instead of freezing your game. They were supposed to do this before, but clay wrote the code wrong and none of us caught it until now. (Fixes #812)

### CODE REVIEW

**Comment by JohnCorby** [comment]
i would like to comment this a ton before merging

File: NewHorizons/Utility/Files/SlideReelAsyncImageLoader.cs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    public ImageLoadedEvent imageLoadedEvent = new();
+
+    private readonly object _lockObj = new();
+
+    public bool FinishedLoading { get; private set; }
+    private int _loadedCount = 0;
+
+    // TODO: set up an optional “StartLoading” and “StartUnloading” condition on AsyncTextureLoader,
+    // and make use of that for at least for projector stuff (require player to be in the same sector as the slides
+    // for them to start loading, and unload when the player leaves)
```
Comment (JohnCorby) [inline NewHorizons/Utility/Files/SlideReelAsyncImageLoader.cs:31]: gotta remember ship log bs too. woohoo more patching!

File: NewHorizons/Builder/Props/ProjectionBuilder.cs [hunk 2]
```diff
@@ -438,17 +477,28 @@ private static ImageUtilities.AsyncImageLoader AddAsyncLoader(GameObject gameObj
...
                 {
-                    imageLoader.imageLoadedEvent?.Invoke(Texture2D.blackTexture, i);
+                    imageLoader.imageLoadedEvent?.Invoke(Texture2D.blackTexture, i, null);
                 }
                 else
                 {
+                    invertedImageLoader.PathsToLoad.Add((i, Path.Combine(mod.ModHelper.Manifest.ModFolderPath, INVERTED_SLIDE_CACHE_FOLDER, slideInfo.imagePath)));
+                    invertedImageLoader.imageLoadedEvent.AddListener((Texture2D t, int i, string s) => 
+                    {
+                        var path = "\\" + Path.Combine(Path.GetFileName(Path.GetDirectoryName(mod.ModHelper.Manifest.ModFolderPath)), slides[i].imagePath) + " > invert";
```
Comment (JohnCorby) [inline NewHorizons/Builder/Props/ProjectionBuilder.cs:487]: i want to change GetKey to not have leading slash

File: NewHorizons/Utility/Files/ImageUtilities.cs [hunk 3]
```diff
@@ -17,14 +15,14 @@ public static class ImageUtilities
...
+        public static void TrackCachedTexture(string key, Texture texture) => _textureCache[key] = texture;
 
-        private static string GetKey(string path) => path.Substring(Main.Instance.ModHelper.OwmlConfig.ModsPath.Length);
+        public static string GetKey(string path) => path.Substring(Main.Instance.ModHelper.OwmlConfig.ModsPath.Length);
 
         public static bool IsTextureLoaded(IModBehaviour mod, string filename)
         {
             var path = Path.Combine(mod.ModHelper.Manifest.ModFolderPath, filename);
-            var key = GetKey(path);
+            var key = GetKey(path).Replace('\\', '/');
```
Comment (JohnCorby) [inline NewHorizons/Utility/Files/ImageUtilities.cs:25]: this replace should probably be in getkey so it replaces \ with / all the time

**Comment by JohnCorby** [comment]
having pikpik test it and then we merge yay

**Review by xen-42 — APPROVED** [review]
Tested on Eyes of the Past, The Machine, and Divergence. Tested loop 1 and loop 2 and a reload after caching and all worked.

**Pull request merged**